include/picongpu/particles/collision/detail/ListEntry.hpp

```diff
@@ -0,0 +1,292 @@
+
+namespace picongpu
+{
+    namespace particles
+    {
+        namespace collision
+        {
+            namespace detail
+            {
+                /* Storage for particle IDs.
+                 *
+                 * Storage for IDs of collision eligible macro particles. It comes with a simple
+                 * shuffling algorithm.
+                 */
+                struct ListEntry
+                {
+                    //! Size of the particle list (number of stored IDs).
+                    uint32_t size;
+                    //! Pointer to the actual data stored on the device heap.
+                    uint32_t* ptrToIndicies;
+
+                    /* Initialize storage, allocate memory.
+                     *
+                     * @param acc alpaka accelerator
+                     * @param deviceHeapHandle Heap handle used for allocating storage on device.
+                     * @param numPar maximal number of IDs to be stored in the list.
+                     */
+                    template<typename T_acc, typename T_DeviceHeapHandle>
+                    DINLINE void init(T_acc const& acc, T_DeviceHeapHandle& deviceHeapHandle, uint32_t numPar)
+                    {
+                        ptrToIndicies = nullptr;
+                        if(numPar != 0u)
+                        {
+                            const int maxTries = 13; // magic number is not performance critical
+                            for(int numTries = 0; numTries < maxTries; ++numTries)
+                            {
+#if(BOOST_LANG_CUDA || BOOST_COMP_HIP) // Allocate memory on a GPU device
+                                ptrToIndicies = (uint32_t*) deviceHeapHandle.malloc(acc, sizeof(uint32_t) * numPar);
+#else // No cuda or hip means the device heap is the host heap.
+                                ptrToIndicies = new uint32_t[numPar];
+#endif
+                                if(ptrToIndicies != nullptr)
+                                {
+                                    break;
+                                }
+                                else
+                                {
+#ifndef BOOST_COMP_HIP
+                                    // TODO: took it from ParticleBox.hpp but HIP should support printf, though
+                                    // maby only with HCC_ENABLE_PRINTF set.
+                                    printf(
+                                        "%s in collisions:  mallocMC out of memory (try %i of %i)\n",
+                                        (numTries + 1) == maxTries ? "ERROR" : "WARNING",
+                                        numTries + 1,
+                                        maxTries);
+#endif
+                                }
+                            }
+                            if(ptrToIndicies == nullptr)
+                            {
+#if BOOST_COMP_HIP
+                                // do nothing
+#elif BOOST_LANG_CUDA
+                                __trap();
```

Comment:
maybe use the new device assert

Reply:
but that would be triggered only when compiling without NDEBUG? That's not the usual case.  But I could just use `assert(0);` directly since it's supported by both HIP and CUDA.

Reply:
OK let us keep it for now until we have a better error notification mechanism in PIConGPU.